### ISSUE TITLE
Using jacfwd/jacrev on stacks of tensors

### ISSUE THREAD

**Issue by bhaveshshrimali** [google/jax#4127]
Hi, 

I am trying to take derivatives with respect to stacks of square matrices, for e.g. `A = (M,M,...,...)` and using `jacrev`/`jacfwd`. For instance consider the simplified case when `A` is a `3x3` matrix, and the function whose derivative I am interested in taking is `A_ij A_ij` (einstein summation convention !) or in other words ` trace (A.T.dot(A))`. The analytical derivative is `2*A`. 

When `A` is a square matrix, I get it perfectly as expected

```python3
import jax.numpy as jnp
from jax import jacfwd, jacrev

def psivec(F):
    I1 = jnp.einsum("ij...,ij...", F, F)
    return I1

def resvecrev(F):
    return jacrev(psivec)(F)

stat = np.random.random((3,3))
F = 0.5 * (np.einsum("ij...->ji...",stat) + stat)
assert(abs(resvecrev(F)-2*F).max() < 1.e-8)
```

However when `A` is several `3x3` matrices stacked together. In other words, say an array of shape `(3,3,100)`, the output of `jacrev(psivec)(F)` is of shape `(100,3,3,100)` which I don't fully understand. 

If I sum along the first axis, i.e. `resvecrev(F).sum(axis=0)`, it does seem to give me what I expect, namely

```python3
print(abs(resvecrev(F).sum(axis=0)-2*F).max() < 1.e-8)   #True
```

Similarly for even higher dimensional stacks say `A.shape = (3,3,100,5)` the output is shaped `(100,5,3,3,100,5)`. Again, summing along the first two axis gives the desired result but I would like to know how to make sense of the output shape and if there is a more efficient way of doing this.  

The [documentation](https://jax.readthedocs.io/en/latest/_modules/jax/api.html#jacrev) is pretty sparse on what to expect for such cases.

**Comment by shoyer** [comment]
It sounds like you want to calculate a Jacobian with vectorization over the trailing axes. You can do this by combining `jacrev` with `vmap`:
```python
import jax.numpy as jnp
import numpy as np
from jax import jacrev, vmap

def psivec(F):
    I1 = jnp.einsum("ij...,ij...", F, F)
    return I1

stat = np.random.random((3,3,10))
F = 0.5 * (np.einsum("ij...->ji...",stat) + stat)
J = vmap(jacrev(psivec), in_axes=(2,), out_axes=2)(F)
assert J.shape == (3, 3, 10)
np.testing.assert_allclose(2*F, J)
```

Note: you might find it a little more ergonomic to put trailing indices at the start. That matches the general NumPy convention, so you can use utilities like `jnp.vectorize` to handle any number of leading dimensions, e.g.,
```python
import jax.numpy as jnp
import numpy as np
from jax import jacrev, vmap

def psivec(F):
    I1 = jnp.einsum("...ij,...ij", F, F)
    return I1

stat = np.random.random((10, 20, 3, 3))
F = 0.5 * (np.einsum("...ij->...ji",stat) + stat)
J = np.vectorize(jacrev(psivec), signature='(i,j)->(i,j)')(F)
assert J.shape == (10, 20, 3, 3,)
np.testing.assert_allclose(2*F, J)
```

It out that `jacfwd`/`jacrev` themselves are constructed from the combination of `vmap` and `vjp`/`jvp`. See the [autodiff cookbook](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Composing-VJPs,-JVPs,-and-vmap) for more details.

**Comment by bhaveshshrimali** [comment]
Thanks @shoyer 
This makes it clearer
